lib/graphql/schema/build_from_definition.rb

```diff
@@ -87,7 +87,13 @@ def build(document, default_resolve: DefaultResolve)
             subscription_root_type = types['Subscription']
           end
 
-          raise InvalidDocumentError.new('Must provide schema definition with query type or a type named Query.') unless query_root_type
+          unless query_root_type
+            raise InvalidDocumentError.new('Must provide schema definition with query type or a type named Query.')
+          end
+
+          unless query_root_type.fields.any?
```

Comment:
This just handles here but wouldn't catch defining the schema normally (via class). You could add a new schema validation to handle both cases: https://github.com/rmosolgo/graphql-ruby/blob/40c838eb4a9f95b9517e7c98077e3cd294ab4677/lib/graphql/schema/validation.rb